File: mobile/ios/OpenU60/Core/Models/SignalModels.swift
import Foundation

struct NRSignal: Equatable {
    var rsrp: Double?
    var rsrq: Double?
    var sinr: Double?
    var rssi: Double?
    var band: String = ""
    var pci: String = ""
    var cellID: String = ""
    var channel: String = ""
    var bandwidth: String = ""
    var carrierAggregation: String = ""
    var sccCarriers: [LTECarrier] = []

    static let empty = NRSignal()

    var isConnected: Bool { rsrp != nil }

    var hasSignal: Bool {
        isConnected || sccCarriers.contains(where: { $0.rsrp != nil })
    }
}

struct LTECarrier: Equatable, Identifiable {
    var id: String { "\(label)-\(band)-\(pci)-\(earfcn)" }
    var label: String = ""
    var pci: String = ""
    var band: String = ""
    var earfcn: String = ""
    var bandwidth: String = ""
    var rsrp: Double?
    var rsrq: Double?
    var sinr: Double?
    var rssi: Double?
}

struct LTESignal: Equatable {
    var rsrp: Double?
    var rsrq: Double?
    var sinr: Double?
    var rssi: Double?
    var pci: String = ""
    var band: String = ""
    var earfcn: String = ""
    var bandwidth: String = ""
    var cellID: String = ""
    var carrierAggregation: String = ""
    var caState: String = ""
    var sccCarriers: [LTECarrier] = []

    static let empty = LTESignal()

    var isConnected: Bool { rsrp != nil }

    var hasSignal: Bool {
        isConnected || sccCarriers.contains(where: { $0.rsrp != nil })
    }
}

struct WCDMASignal: Equatable {
    var rscp: Double?
    var ecio: Double?

    static let empty = WCDMASignal()

    var isConnected: Bool { rscp != nil }
}

struct OperatorInfo: Equatable {
    var provider: String = ""
    var networkType: String = ""
    var signalBar: Int = 0
    var roaming: Bool = false

    static let empty = OperatorInfo()

    enum NetworkMode: Equatable {
        case sa, nsa, lte, legacy, unknown
    }

    var networkMode: NetworkMode {
        let raw = networkType.uppercased()
        if raw == "SA" || raw.contains("5G SA") || raw.contains("NR SA") || raw.contains("NR-SA") { return .sa }
        if raw.contains("NSA") || raw == "ENDC" || raw == "EN-DC" { return .nsa }
        if raw.contains("LTE") || raw == "4G" || raw == "4G+" { return .lte }
        if raw.contains("WCDMA") || raw.contains("UMTS") || raw.contains("GSM")
            || raw.contains("2G") || raw.contains("3G") { return .legacy }
        return .unknown
    }

    func displayNetworkType(nrConnected: Bool, lteSignal: LTESignal = .empty) -> String {
        // Firmware says LTE but NR is actually connected → 5G NSA
        if nrConnected && (networkMode == .lte || networkMode == .unknown) {
            return "5G NSA"
        }
        // Firmware still says SA/NSA but NR has dropped → fall back to 4G
        if !nrConnected && (networkMode == .sa || networkMode == .nsa) {
            if lteSignal.isConnected {
                return lteSignal.sccCarriers.isEmpty ? "4G" : "4G+"
            }
            return "4G"
        }
        switch networkMode {
        case .sa: return "5G SA"
        case .nsa: return "5G NSA"
        case .lte:
            let raw = networkType.uppercased()
            return (raw.contains("CA") || raw == "4G+" || raw.contains("LTE-A") || raw.contains("LTE+"))
                ? "4G+" : "4G"
        case .legacy: return networkType
        case .unknown: return networkType
        }
    }

    func showNR(nr: NRSignal) -> Bool {
        nr.hasSignal
    }

    func showLTE(lte: LTESignal) -> Bool {
        if networkMode == .sa { return false }
        let raw = networkType.uppercased()
        let hasData = lte.hasSignal
        let actHintsLTE = raw.contains("NSA") || raw.contains("LTE") || raw.contains("E-UTRAN")
            || raw.contains("ENDC") || raw.contains("EN-DC") || raw == "4G" || raw == "4G+"
        let actHintsNR = raw.contains("SA") || raw.contains("NR") || raw.contains("5G")
            || raw.contains("ENDC") || raw.contains("EN-DC")
        return hasData && (actHintsLTE || raw.isEmpty || actHintsNR)
    }

    func show3G(nr: NRSignal, lte: LTESignal, wcdma: WCDMASignal) -> Bool {
        !showNR(nr: nr) && !showLTE(lte: lte) && (wcdma.rscp != nil || wcdma.ecio != nil)
    }
}

struct SignalSnapshot: Identifiable, Equatable {
    let id = UUID()
    let timestamp: Date
    let nrRSRP: Double?
    let lteRSRP: Double?
    let wcdmaRSCP: Double?

    init(timestamp: Date, nrRSRP: Double?, lteRSRP: Double?, wcdmaRSCP: Double? = nil) {
        self.timestamp = timestamp
        self.nrRSRP = nrRSRP
        self.lteRSRP = lteRSRP
        self.wcdmaRSCP = wcdmaRSCP
    }
}

/// Parser that extracts signal data from the agent nwinfo_get_netinfo response.
enum SignalParser {

    /// One entry of an `lteca` / `nrca` list: "PCI,Band,Index,EARFCN,BW".
    private typealias CarrierEntry = (pci: String, band: String, earfcn: String, bandwidth: String)

    /// One entry of an `ltecasig` / `nrcasig` list: "RSRP,RSRQ,SINR,RSSI,...".
    private typealias CarrierSignal = (rsrp: Double?, rsrq: Double?, sinr: Double?, rssi: Double?)

    static func parseNetInfo(_ data: [String: Any]) -> (NRSignal, LTESignal, WCDMASignal, OperatorInfo) {
        var nr = NRSignal()
        var lte = LTESignal()
        var wcdma = WCDMASignal()
        var op = OperatorInfo()

        let nrcaStr = stringVal(data["nrca"])

        nr.rsrp = parseDouble(data["nr5g_rsrp"]).flatMap { $0 == 0 ? nil : $0 }
        nr.rsrq = parseDouble(data["nr5g_rsrq"])
        nr.sinr = parseDouble(data["nr5g_snr"])
        nr.rssi = parseDouble(data["nr5g_rssi"])
        nr.band = stringVal(data["nr5g_action_band"])
        nr.pci = stringVal(data["nr5g_pci"])
        nr.cellID = stringVal(data["nr5g_cell_id"])
        nr.channel = stringVal(data["nr5g_action_channel"])
        nr.bandwidth = stringVal(data["nr5g_bandwidth"])
        nr.carrierAggregation = nrcaStr

        // nrca/nrcasig use the same wire format as lteca/ltecasig.
        let nrSplit = splitPCC(parseCarriers(nrcaStr), pci: nr.pci, channel: nr.channel)
        if let pccBandwidth = nrSplit.pccBandwidth { nr.bandwidth = pccBandwidth }
        nr.sccCarriers = makeCarriers(
            nrSplit.sccs,
            signals: parseCarrierSignals(stringVal(data["nrcasig"])),
            labelPrefix: "5G SCC"
        )

        let pccPci = stringVal(data["lte_pci"])
        let pccEarfcn = stringVal(data["wan_active_channel"])
        lte.rsrp = parseDouble(data["lte_rsrp"]).flatMap { $0 == 0 ? nil : $0 }
        lte.rsrq = parseDouble(data["lte_rsrq"])
        lte.sinr = parseDouble(data["lte_snr"])
        lte.rssi = parseDouble(data["lte_rssi"])
        lte.pci = pccPci
        lte.earfcn = pccEarfcn
        lte.band = stringVal(data["wan_active_band"])
        lte.cellID = stringVal(data["cell_id"])
        lte.caState = stringVal(data["lteca_state"])

        let ltecaStr = stringVal(data["lteca"])
        lte.carrierAggregation = ltecaStr

        let lteSplit = splitPCC(parseCarriers(ltecaStr), pci: pccPci, channel: pccEarfcn)
        if let pccBandwidth = lteSplit.pccBandwidth { lte.bandwidth = pccBandwidth }
        lte.sccCarriers = makeCarriers(
            lteSplit.sccs,
            signals: parseCarrierSignals(stringVal(data["ltecasig"])),
            labelPrefix: "SCC"
        )

        wcdma.rscp = parseDouble(data["rscp"]).flatMap { $0 == 0 ? nil : $0 }
        wcdma.ecio = parseDouble(data["ecio"])

        op.provider = stringVal(data["network_provider"])
        op.networkType = stringVal(data["network_type"])
        op.signalBar = Int(stringVal(data["signalbar"])) ?? 0
        op.roaming = stringVal(data["simcard_roam"]) == "1"

        return (nr, lte, wcdma, op)
    }

    // MARK: - Carrier lists

    /// `split` already omits empty subsequences, so leading/trailing/repeated separators need
    /// no pre-trimming.
    private static func parseCarriers(_ raw: String) -> [CarrierEntry] {
        var entries: [CarrierEntry] = []
        for entry in raw.split(separator: ";") {
            let parts = entry.split(separator: ",")
            guard parts.count >= 5 else { continue }
            entries.append((
                pci: String(parts[0]),
                band: String(parts[1]),
                earfcn: String(parts[3]),
                bandwidth: String(parts[4])
            ))
        }
        return entries
    }

    private static func parseCarrierSignals(_ raw: String) -> [CarrierSignal] {
        var signals: [CarrierSignal] = []
        for entry in raw.split(separator: ";") {
            let parts = entry.split(separator: ",")
            guard parts.count >= 4 else { continue }
            signals.append((
                rsrp: parseField(parts[0]),
                rsrq: parseField(parts[1]),
                sinr: parseField(parts[2]),
                rssi: parseField(parts[3])
            ))
        }
        return signals
    }

    /// Separates the primary carrier — matched on PCI plus channel — from the secondaries.
    /// - Returns: the PCC's own bandwidth when the firmware reported one, and the rest.
    private static func splitPCC(_ carriers: [CarrierEntry], pci: String, channel: String)
        -> (pccBandwidth: String?, sccs: [CarrierEntry]) {
        var pccBandwidth: String?
        var sccs: [CarrierEntry] = []
        var pccFound = false
        for carrier in carriers {
            if !pccFound && !pci.isEmpty && carrier.pci == pci && carrier.earfcn == channel {
                if !carrier.bandwidth.isEmpty { pccBandwidth = carrier.bandwidth }
                pccFound = true
            } else {
                sccs.append(carrier)
            }
        }
        return (pccBandwidth, sccs)
    }

    private static func makeCarriers(_ entries: [CarrierEntry], signals: [CarrierSignal], labelPrefix: String) -> [LTECarrier] {
        entries.enumerated().map { index, entry in
            var carrier = LTECarrier(
                label: "\(labelPrefix)\(index)",
                pci: entry.pci,
                band: entry.band,
                earfcn: entry.earfcn,
                bandwidth: entry.bandwidth
            )
            if index < signals.count {
                carrier.rsrp = signals[index].rsrp
                carrier.rsrq = signals[index].rsrq
                carrier.sinr = signals[index].sinr
                carrier.rssi = signals[index].rssi
            }
            return carrier
        }
    }

    // MARK: - Scalars

    /// Parses one CSV field in place, without materialising a `String` for it.
    private static func parseField(_ field: Substring) -> Double? {
        var slice = field
        while let first = slice.first, first.isWhitespace { slice.removeFirst() }
        while let last = slice.last, last.isWhitespace { slice.removeLast() }
        return Double(slice)
    }

    private static func parseDouble(_ value: Any?) -> Double? {
        var result: Double?
        if let d = value as? Double { result = d }
        else if let i = value as? Int { result = Double(i) }
        else if let s = value as? String {
            let trimmed = s.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty || trimmed == "--" || trimmed == "N/A" { return nil }
            result = Double(trimmed)
        }
        if let r = result, r > 9000 || r < -9000 { return nil }
        return result
    }

    private static func stringVal(_ value: Any?) -> String {
        if let s = value as? String { return s }
        if let i = value as? Int { return String(i) }
        if let d = value as? Double { return String(d) }
        return ""
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/core/model/SignalModels.kt
package com.openu60.core.model

data class LTECarrier(
    val label: String = "",
    val pci: String = "",
    val band: String = "",
    val earfcn: String = "",
    val bandwidth: String = "",
    val rsrp: Double? = null,
    val rsrq: Double? = null,
    val sinr: Double? = null,
    val rssi: Double? = null,
) {
    val id: String get() = "$label-$band-$pci-$earfcn"
}

data class NRSignal(
    val rsrp: Double? = null,
    val rsrq: Double? = null,
    val sinr: Double? = null,
    val rssi: Double? = null,
    val band: String = "",
    val pci: String = "",
    val cellID: String = "",
    val channel: String = "",
    val bandwidth: String = "",
    val carrierAggregation: String = "",
    val sccCarriers: List<LTECarrier> = emptyList(),
) {
    val isConnected: Boolean get() = rsrp != null
    val hasSignal: Boolean get() = isConnected || sccCarriers.any { it.rsrp != null }

    companion object {
        val empty = NRSignal()
    }
}

data class LTESignal(
    val rsrp: Double? = null,
    val rsrq: Double? = null,
    val sinr: Double? = null,
    val rssi: Double? = null,
    val pci: String = "",
    val band: String = "",
    val earfcn: String = "",
    val bandwidth: String = "",
    val cellID: String = "",
    val carrierAggregation: String = "",
    val caState: String = "",
    val sccCarriers: List<LTECarrier> = emptyList(),
) {
    val isConnected: Boolean get() = rsrp != null
    val hasSignal: Boolean get() = isConnected || sccCarriers.any { it.rsrp != null }

    companion object {
        val empty = LTESignal()
    }
}

data class WCDMASignal(
    val rscp: Double? = null,
    val ecio: Double? = null,
) {
    val isConnected: Boolean get() = rscp != null

    companion object {
        val empty = WCDMASignal()
    }
}

data class OperatorInfo(
    val provider: String = "",
    val networkType: String = "",
    val signalBar: Int = 0,
    val roaming: Boolean = false,
) {
    enum class NetworkMode { SA, NSA, LTE, LEGACY, UNKNOWN }

    val networkMode: NetworkMode
        get() {
            val raw = networkType.uppercase()
            return when {
                raw == "SA" || raw.contains("5G SA") || raw.contains("NR SA") || raw.contains("NR-SA") -> NetworkMode.SA
                raw.contains("NSA") || raw == "ENDC" || raw == "EN-DC" -> NetworkMode.NSA
                raw.contains("LTE") || raw == "4G" || raw == "4G+" -> NetworkMode.LTE
                raw.contains("WCDMA") || raw.contains("UMTS") || raw.contains("GSM")
                    || raw.contains("2G") || raw.contains("3G") -> NetworkMode.LEGACY
                else -> NetworkMode.UNKNOWN
            }
        }

    fun displayNetworkType(nrConnected: Boolean, lteSignal: LTESignal = LTESignal.empty): String {
        if (nrConnected && (networkMode == NetworkMode.LTE || networkMode == NetworkMode.UNKNOWN)) {
            return "5G NSA"
        }
        if (!nrConnected && (networkMode == NetworkMode.SA || networkMode == NetworkMode.NSA)) {
            return if (lteSignal.isConnected) {
                if (lteSignal.sccCarriers.isEmpty()) "4G" else "4G+"
            } else "4G"
        }
        return when (networkMode) {
            NetworkMode.SA -> "5G SA"
            NetworkMode.NSA -> "5G NSA"
            NetworkMode.LTE -> {
                val raw = networkType.uppercase()
                if (raw.contains("CA") || raw == "4G+" || raw.contains("LTE-A") || raw.contains("LTE+")) "4G+" else "4G"
            }
            NetworkMode.LEGACY -> networkType
            NetworkMode.UNKNOWN -> networkType
        }
    }

    fun showNR(nr: NRSignal): Boolean = nr.hasSignal

    fun showLTE(lte: LTESignal): Boolean {
        if (networkMode == NetworkMode.SA) return false
        val raw = networkType.uppercase()
        val hasData = lte.hasSignal
        val actHintsLTE = raw.contains("NSA") || raw.contains("LTE") || raw.contains("E-UTRAN")
            || raw.contains("ENDC") || raw.contains("EN-DC") || raw == "4G" || raw == "4G+"
        val actHintsNR = raw.contains("SA") || raw.contains("NR") || raw.contains("5G")
            || raw.contains("ENDC") || raw.contains("EN-DC")
        return hasData && (actHintsLTE || raw.isEmpty() || actHintsNR)
    }

    fun show3G(nr: NRSignal, lte: LTESignal, wcdma: WCDMASignal): Boolean {
        return !showNR(nr) && !showLTE(lte) && (wcdma.rscp != null || wcdma.ecio != null)
    }

    companion object {
        val empty = OperatorInfo()
    }
}

data class SignalSnapshot(
    val id: Int,
    val timestamp: Long,
    val nrRSRP: Double?,
    val lteRSRP: Double?,
) {
    companion object {
        private var nextID = 0
        fun create(nrRSRP: Double?, lteRSRP: Double?): SignalSnapshot {
            return SignalSnapshot(
                id = nextID++,
                timestamp = System.currentTimeMillis(),
                nrRSRP = nrRSRP,
                lteRSRP = lteRSRP,
            )
        }
    }
}

// MARK: - Parser

object SignalParser {
    fun parseNetInfo(data: Map<String, Any?>): SignalResult {
        var nr = NRSignal.empty
        var lte = LTESignal.empty
        var wcdma = WCDMASignal.empty
        var op = OperatorInfo.empty

        nr = nr.copy(
            rsrp = parseSignalDouble(data["nr5g_rsrp"]),
            rsrq = parseSignalDouble(data["nr5g_rsrq"]),
            sinr = parseSignalDouble(data["nr5g_snr"]),
            rssi = parseSignalDouble(data["nr5g_rssi"]),
            band = stringVal(data["nr5g_action_band"]),
            pci = stringVal(data["nr5g_pci"]),
            cellID = stringVal(data["nr5g_cell_id"]),
            channel = stringVal(data["nr5g_action_channel"]),
            bandwidth = stringVal(data["nr5g_bandwidth"]),
            carrierAggregation = stringVal(data["nrca"]),
        )

        // Parse nrca: "PCI,Band,Index,EARFCN,BW;..."
        val nrcaStr = stringVal(data["nrca"])
        val nrCarriers = parseCAString(nrcaStr)
        val nrcasigStr = stringVal(data["nrcasig"])
        val nrSccSigs = parseCASigString(nrcasigStr)

        // Match NR PCC by PCI+channel
        val nrPccPci = nr.pci
        val nrPccChannel = nr.channel
        val nrSccEntries = mutableListOf<CarrierEntry>()
        var nrPccFound = false
        var updatedNrBw = nr.bandwidth
        for (c in nrCarriers) {
            if (!nrPccFound && c.pci == nrPccPci && c.earfcn == nrPccChannel && nrPccPci.isNotEmpty()) {
                if (c.bandwidth.isNotEmpty()) updatedNrBw = c.bandwidth
                nrPccFound = true
            } else {
                nrSccEntries.add(c)
            }
        }
        nr = nr.copy(bandwidth = updatedNrBw)

        val nrSccCarriers = nrSccEntries.mapIndexed { i, sc ->
            val sig = nrSccSigs.getOrNull(i)
            LTECarrier(
                label = "5G SCC$i", pci = sc.pci, band = sc.band,
                earfcn = sc.earfcn, bandwidth = sc.bandwidth,
                rsrp = sig?.rsrp, rsrq = sig?.rsrq, sinr = sig?.sinr, rssi = sig?.rssi,
            )
        }
        nr = nr.copy(sccCarriers = nrSccCarriers)

        // LTE
        val pccPci = stringVal(data["lte_pci"])
        val pccEarfcn = stringVal(data["wan_active_channel"])
        lte = lte.copy(
            rsrp = parseSignalDouble(data["lte_rsrp"]),
            rsrq = parseSignalDouble(data["lte_rsrq"]),
            sinr = parseSignalDouble(data["lte_snr"]),
            rssi = parseSignalDouble(data["lte_rssi"]),
            pci = pccPci,
            earfcn = pccEarfcn,
            band = stringVal(data["wan_active_band"]),
            cellID = stringVal(data["cell_id"]),
            caState = stringVal(data["lteca_state"]),
        )

        val ltecaStr = stringVal(data["lteca"])
        lte = lte.copy(carrierAggregation = ltecaStr)
        val carriers = parseCAString(ltecaStr)
        val ltecasigStr = stringVal(data["ltecasig"])
        val sccSigs = parseCASigString(ltecasigStr)

        val sccEntries = mutableListOf<CarrierEntry>()
        var pccFound = false
        var updatedLteBw = lte.bandwidth
        for (c in carriers) {
            if (!pccFound && c.pci == pccPci && c.earfcn == pccEarfcn && pccPci.isNotEmpty()) {
                if (c.bandwidth.isNotEmpty()) updatedLteBw = c.bandwidth
                pccFound = true
            } else {
                sccEntries.add(c)
            }
        }
        lte = lte.copy(bandwidth = updatedLteBw)

        val sccCarriers = sccEntries.mapIndexed { i, sc ->
            val sig = sccSigs.getOrNull(i)
            LTECarrier(
                label = "SCC$i", pci = sc.pci, band = sc.band,
                earfcn = sc.earfcn, bandwidth = sc.bandwidth,
                rsrp = sig?.rsrp, rsrq = sig?.rsrq, sinr = sig?.sinr, rssi = sig?.rssi,
            )
        }
        lte = lte.copy(sccCarriers = sccCarriers)

        wcdma = wcdma.copy(
            rscp = parseSignalDouble(data["rscp"]),
            ecio = parseSignalDouble(data["ecio"]),
        )

        op = op.copy(
            provider = stringVal(data["network_provider"]),
            networkType = stringVal(data["network_type"]),
            signalBar = stringVal(data["signalbar"]).toIntOrNull() ?: 0,
            roaming = stringVal(data["simcard_roam"]) == "1",
        )

        return SignalResult(nr, lte, wcdma, op)
    }

    private data class CarrierEntry(val pci: String, val band: String, val earfcn: String, val bandwidth: String)
    private data class SigEntry(val rsrp: Double?, val rsrq: Double?, val sinr: Double?, val rssi: Double?)

    private fun parseCAString(str: String): List<CarrierEntry> {
        if (str.isBlank()) return emptyList()
        return str.trimEnd(';').split(";").mapNotNull { entry ->
            val parts = entry.split(",")
            if (parts.size >= 5) CarrierEntry(parts[0], parts[1], parts[3], parts[4])
            else null
        }
    }

    private fun parseCASigString(str: String): List<SigEntry> {
        if (str.isBlank()) return emptyList()
        return str.trimEnd(';').split(";").mapNotNull { entry ->
            val parts = entry.split(",")
            if (parts.size >= 4) SigEntry(
                parts[0].trim().toDoubleOrNull(),
                parts[1].trim().toDoubleOrNull(),
                parts[2].trim().toDoubleOrNull(),
                parts[3].trim().toDoubleOrNull(),
            ) else null
        }
    }

    private fun parseSignalDouble(value: Any?): Double? {
        val result = when (value) {
            is Double -> value
            is Int -> value.toDouble()
            is String -> {
                val trimmed = value.trim()
                if (trimmed.isEmpty() || trimmed == "--" || trimmed == "N/A") null
                else trimmed.toDoubleOrNull()
            }
            else -> null
        }
        return result?.takeIf { it in -9000.0..9000.0 }?.takeIf { it != 0.0 }
    }

    private fun stringVal(value: Any?): String = when (value) {
        is String -> value
        is Int -> value.toString()
        is Double -> value.toString()
        else -> ""
    }
}

data class SignalResult(
    val nr: NRSignal,
    val lte: LTESignal,
    val wcdma: WCDMASignal,
    val operatorInfo: OperatorInfo,
)
